kernel_modules/kmod_meltdown/meltdown_module.c

```diff
@@ -7,96 +7,95 @@
  * SPDX-License-Identifier: BSD-3-Clause OR GPL-2.0
  */
 
+#include <linux/debugfs.h>
 #include <linux/module.h>
-#include <linux/proc_fs.h>
-#include <linux/seq_file.h>
 
 MODULE_LICENSE("Dual BSD/GPL");
 MODULE_AUTHOR("Google");
 MODULE_DESCRIPTION("");
-MODULE_VERSION("0.1");
-
-// Keeps a buffer in kernel memory and makes its address available at:
-//   /proc/safeside_meltdown/address
-// and its length available at:
-//   /proc/safeside_meltdown/length
-// which are both only accessible to root.
-
-// Secret data stored in the kernel memory whose content is never directly
-// leaked through sysfs.
-const char *private_data = "It's a s3kr3t!!!";
-
-// Directory record. Must be available on unloading the module.
-struct proc_dir_entry *safeside_meltdown;
-
-// Print the address of `private_data` into the provided buffer.
-static int address_show(struct seq_file *file, void *v) {
-  seq_printf(file, "%px\n", private_data);
-  return 0;
-}
+MODULE_VERSION("0.2");
+
+// This module provides the kernel side of a test for Meltdown.
+//
+// It holds a buffer of "secret" data in kernel memory that is never made
+// (legitimately) available to userspace. Through debugfs, userspace can find
+// the address and length of that buffer and can force it to be read into
+// cache.
+//
+// Assuming debugfs is mounted in the usual place, the files are at:
+//   /sys/kernel/debug/safeside_meltdown/
+// And they are:
+//   secret_data_address
+//     virtual address of secret data, returned as "0x123"
+//   secret_data_length
+//     length of secret data in bytes, returned as "123"
+//   secret_data_in_cache
+//     when opened and read, forces secret data to be read into cache.
+//
+// Access is restricted to root to avoid opening any new attack surface, e.g.
```

Comment:
This implies that it's hypothetical, but the values of `secret_data_address` and `secret_data_length` are  specifically writable, so the vulnerability is a bit more realistic, right? In particular anyone with write access can change the term of the bounds check in `secret_data_in_cache_get`, which is meant to be a gadget to begin with.

This is the only thing that really concerns me in this review. If I understand correctly, it's a big change from the "before" version, where it was read-only. As long as we're OK with this and it's clearly documented, LGTM. Rest of the comments are optional / FYI.

Reply:
Thankfully these *aren't* writable because of the mode we pass to `debugfs_create_u32`:

```c
  child = debugfs_create_u32("secret_data_length", 0400, debugfs_dir,
      &secret_data_length);
```

`0400` is read-only. Likewise for `secret_data_address`.

Reply:
Ah, I didn't catch that. Subtle subtle subtle. Thanks!